src/include/txn.i

```diff
@@ -727,29 +723,11 @@ __wt_txn_read(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_UPDATE *upd, WT
             if (upd_visible == WT_VISIBLE_PREPARE)
                 return (WT_PREPARE_CONFLICT);
         }
-        /* An invisible birthmark is equivalent to a tombstone. */
-        if (upd->type == WT_UPDATE_BIRTHMARK)
-            skipped_birthmark = true;
-    }
-
-    /* Find a visible record in the lookaside and check if we should use that instead. */
-    if ((upd == NULL || session->txn.read_timestamp != 0) &&
-      __wt_page_las_active(session, cbt->ref)) {
-        ret = __wt_find_lookaside_upd(session, cbt, &tmp_upd, false);
-        WT_RET_NOTFOUND_OK(ret);
-        if (ret == 0) {
-            /* Pick newer of the visible updates */
-            if (upd == NULL || upd->start_ts < tmp_upd->start_ts)
-                upd = tmp_upd;
-            else {
-                WT_STAT_CONN_INCR(session, cache_lookaside_read_wasted);
-                __wt_free_update_list(session, &tmp_upd);
-            }
-        }
     }
 
-    if (upd == NULL && skipped_birthmark)
-        upd = &tombstone;
+    /* Cannot find the update in memory. Try lookaside. */
+    if (upd == NULL && __wt_page_las_active(session, cbt->ref))
+        WT_RET_NOTFOUND_OK(__wt_find_lookaside_upd(session, cbt, &upd, false));
```

Comment:
I think the code flow is more confusing (and the `Cannot find the update in memory` comment is now incorrect, as is the function description).

What would you think of something like:
```
/*
 * __wt_txn_read --
 *     Get the first visible update in a list or the history file (or NULL if none are visible).
 */ 
static inline int
__wt_txn_read(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_UPDATE *upd, WT_UPDATE **updp)
{   
    WT_VISIBLE_TYPE upd_visible;
        
    *updp = NULL;

    for (; upd != NULL; upd = upd->next) {
        /* Skip reserved place-holders, they're never visible. */
        if (upd->type == WT_UPDATE_RESERVE)   
                continue;
        upd_visible = __wt_txn_upd_visible_type(session, upd);
        if (upd_visible == WT_VISIBLE_TRUE) {
            if (upd->type == WT_UPDATE_BIRTHMARK)
                *updp = upd;
            return (0);
        }
        if (upd_visible == WT_VISIBLE_PREPARE)
            return (WT_PREPARE_CONFLICT);
    }
        
    /* If there's no visible update in memory, check the history file. */
    if (__wt_page_las_active(session, cbt->ref))
        WT_RET_NOTFOUND_OK(__wt_find_lookaside_upd(session, cbt, &upd, false));
    
    *updp = upd == NULL || upd->type == WT_UPDATE_BIRTHMARK ? NULL : upd;
    return (0);
}   
```

Reply:
Changed as suggested. But I think the line *updp = upd changes the behaviour. It should be upd->type != WT_UPDATE_BIRTHMARK

Reply:
Yes, I agree -- in general, if I don't claim some code suggestion is tested, it almost certainly wasn't. :)

Reply:
Just to confirm :)